File: Yijing.swt/Yijing/src/org/yijing/CLineValueSequencer.java
package org.yijing;

public class CLineValueSequencer extends CValueSequencer
{
	public CLineValueSequencer(int nValue) {
		super(3, 4, nValue);
		m_pvsInner = new CBitValueSequencer[3];
		m_pvsInner[2] = new CBitValueSequencer(0);
		m_pvsInner[1] = new CBitValueSequencer(0);
		m_pvsInner[0] = new CBitValueSequencer(0);
		Bit(2).SetParent(this);
		Bit(1).SetParent(this);
		Bit(0).SetParent(this);
		m_nSequences = Sequences.nLineSequences;
		m_nRatios = Sequences.nLineRatios;
	}

	public CBitValueSequencer Bit(int nIndex) {
		return (CBitValueSequencer)m_pvsInner[nIndex];
	}

	public CValueSequencer Inverse() {
		switch (m_nValue)
		{
			case 0:
				Value(1);
				break;
			case 1:
				Value(0);
				break;
			case 2:
				Value(3);
				break;
			case 3:
				Value(2);
				break;
		}
		UpdateInnerValues();
		UpdateOuterValues();
		return this;
	}

	public CValueSequencer Move() {
		if (Value() == 0)
			Value(1);
		else
			if (Value() == 3)
				Value(2);
		UpdateInnerValues();
		UpdateOuterValues();
		return this;
	}

	public CValueSequencer Young() {
		if (Value() == 0)
			Value(2);
		else
			if (Value() == 3)
				Value(1);
		UpdateInnerValues();
		UpdateOuterValues();
		return this;
	}

	public CValueSequencer Old() {
		if (Value() == 2)
			Value(0);
		else
			if (Value() == 1)
				Value(3);
		UpdateInnerValues();
		UpdateOuterValues();
		return this;
	}

	protected void UpdateInnerValues() {
		switch (m_nValue)
		{
			case 0:
				Bit(2).Value(0);
				Bit(1).Value(0);
				Bit(0).Value(0);
				break;
			case 1:
				Bit(2).Value(0);
				Bit(1).Value(1);
				Bit(0).Value(0);
				break;
			case 2:
				Bit(2).Value(1);
				Bit(1).Value(0);
				Bit(0).Value(1);
				break;
			case 3:
				Bit(2).Value(1);
				Bit(1).Value(1);
				Bit(0).Value(1);
				break;
		}
		Bit(2).UpdateInnerValues();
		Bit(1).UpdateInnerValues();
		Bit(0).UpdateInnerValues();
	}

	protected void UpdateOuterValues() {
		Value(Bit(0).Value() + Bit(1).Value() + Bit(2).Value());
		if(m_pvsParent != null)
			m_pvsParent.UpdateOuterValues();
	}

	public static void SetCurrentSequence(int nSequence) {
		m_nCurrentSequence = nSequence;
	}

	public static void SetCurrentRatio(int nRatio) {
		m_nCurrentRatio = nRatio;
	}
	
	public static void SetCurrentLabel(int nLabel) {
		m_nCurrentLabel = nLabel;
	}
	
	protected String GetLabel() { 
		return Sequences.strLineLabels[GetCurrentLabel()][Value()];
	}
	
	protected boolean GetMoving() {
		return (Value() == 0) || (Value() == 3);
	}

	protected int GetCurrentSequence() { return m_nCurrentSequence; }
	protected int GetCurrentRatio() { return m_nCurrentRatio; }
	protected int GetCurrentLabel() { return m_nCurrentLabel; }

	protected static int m_nCurrentSequence = 0;
	protected static int m_nCurrentRatio = 0;
	protected static int m_nCurrentLabel = 0;
}

File: ValueSequencer.ts/CLineValueSequencer.ts
import { CValueSequencer } from "./CValueSequencer.ts";
import { CBitValueSequencer } from "./CBitValueSequencer.ts";
import { Sequences } from "./Sequences.ts";

export class CLineValueSequencer extends CValueSequencer {
  private static m_nCurrentSequence = 0;
  private static m_nCurrentRatio = 0;
  private static m_nCurrentLabel = 0;

  constructor(nValue: number) {
    super(3, 4, nValue);
    this.m_pvsInner = [
      new CBitValueSequencer(0),
      new CBitValueSequencer(0),
      new CBitValueSequencer(0),
    ];
    this.bit(2).setParent(this);
    this.bit(1).setParent(this);
    this.bit(0).setParent(this);
    this.m_nSequences = Sequences.nLineSequences;
    this.m_nRatios = Sequences.nLineRatios;
  }

  public bit(nIndex: number): CBitValueSequencer {
    return this.m_pvsInner[nIndex] as CBitValueSequencer;
  }

  public override inverse(): this {
    switch (this.m_nValue) {
      case 0:
        this.value = 1;
        break;
      case 1:
        this.value = 0;
        break;
      case 2:
        this.value = 3;
        break;
      case 3:
        this.value = 2;
        break;
    }
    this.updateInnerValues();
    this.updateOuterValues();
    return this;
  }

  public override move(): this {
    if (this.value === 0) {
      this.value = 1;
    } else if (this.value === 3) {
      this.value = 2;
    }
    this.updateInnerValues();
    this.updateOuterValues();
    return this;
  }

  public override young(): this {
    if (this.value === 0) {
      this.value = 2;
    } else if (this.value === 3) {
      this.value = 1;
    }
    this.updateInnerValues();
    this.updateOuterValues();
    return this;
  }

  public override old(): this {
    if (this.value === 2) {
      this.value = 0;
    } else if (this.value === 1) {
      this.value = 3;
    }
    this.updateInnerValues();
    this.updateOuterValues();
    return this;
  }

  public override updateInnerValues(): void {
    switch (this.m_nValue) {
      case 0:
        this.bit(2).value = 0;
        this.bit(1).value = 0;
        this.bit(0).value = 0;
        break;
      case 1:
        this.bit(2).value = 0;
        this.bit(1).value = 1;
        this.bit(0).value = 0;
        break;
      case 2:
        this.bit(2).value = 1;
        this.bit(1).value = 0;
        this.bit(0).value = 1;
        break;
      case 3:
        this.bit(2).value = 1;
        this.bit(1).value = 1;
        this.bit(0).value = 1;
        break;
    }
    this.bit(2).updateInnerValues();
    this.bit(1).updateInnerValues();
    this.bit(0).updateInnerValues();
  }

  public override updateOuterValues(): void {
    this.value = this.bit(0).value + this.bit(1).value + this.bit(2).value;
    this.m_pvsParent?.updateOuterValues();
  }

  public static setCurrentSequence(nSequence: number): void {
    this.m_nCurrentSequence = nSequence;
  }

  public static setCurrentRatio(nRatio: number): void {
    this.m_nCurrentRatio = nRatio;
  }

  public static setCurrentLabel(nLabel: number): void {
    this.m_nCurrentLabel = nLabel;
  }

  protected override getLabel(): string {
    return Sequences.strLineLabels[this.getCurrentLabel()][this.value];
  }

  protected override getMoving(): boolean {
    return this.value === 0 || this.value === 3;
  }

  protected override getCurrentSequence(): number {
    return CLineValueSequencer.m_nCurrentSequence;
  }

  protected override getCurrentRatio(): number {
    return CLineValueSequencer.m_nCurrentRatio;
  }

  protected override getCurrentLabel(): number {
    return CLineValueSequencer.m_nCurrentLabel;
  }
}
